[Dispatcher] Collect autograd sequence numbers on PythonTLSSnapshot dispatch keys

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123304

Fixes #121758

**TL;DR**: When profiling is turned on, the dispatcher will sometimes attach the autograd sequence number to the recorded profiler event. This PR expands the set of profiler events onto which we attach sequence numbers. Before, we'd only attach a sequence number if the current dispatch key was an Autograd dispatch key. Now, we attach a sequence number if the current dispatch key **set** contains Autograd.

**Context**:
The use case for this is torch.profiler for python subclasses.

Autograd attaches a "sequence number" to all ops that it encounters during the forward pass. Then, the corresponding sequence number can be associated with a backward kernel when backward is executed. This is used by the profiler to associate the forward ops to the backward ops; a forward op and a backward op with the same sequence number are "linked" in some post-processing step.

Prior to this PR, this profiler feature didn't work for python subclasses. The reason is that we don't collect profiler information for all the dispatches for a given kernel; we only dispatch the initial `call`, and not the subsequent `redispatch` invocations. Normally, an Autograd key (if we're running with autograd) is the highest dispatch key, so the initial `call` that we profile is an Autograd key, and we collect the sequence number. But when we're dealing with a python subclass, the first dispatch key is PythonTLSSnapshot, which eventually redispatches into Autograd. We don't record the Autograd sequence number in that case because we don't record redispatches.

To fix this, this PR collects a sequence number whenever the dispatch key **set** contains an Autograd key. That means we might sometimes collect multiple events with the same sequence number, or possibly attach sequence numbers when we won't actually use them? (e.g. maybe if the initial dispatch key handler removes Autograd for some reason). Although this might be a bit confusing for users looking directly at the sequence_nr directly, I think the main use case is for the profiler to create fwd-bwd links; and those should still be generated correctly in these cases.

Differential Revision: [D55724190](https://our.internmc.facebook.com/intern/diff/D55724190)

cc @ezyang @msaroufim @albanD

davidberard98 committed:
[Dispatcher] Collect autograd sequence numbers on PythonTLSSnapshot dispatch keys

Fixes #121758

**TL;DR**: When profiling is turned on, the dispatcher will sometimes attach the autograd sequence number to the recorded profiler event. This PR expands the set of profiler events onto which we attach sequence numbers. In addition to the previous events, we will now also attach sequence numbers for an event which is (a) called through "call" (as opposed to "redispatch"), (b) has PythonTLSSnapshot as the current dispatch key, and (c) has an Autograd key in the dispatchKeySet.

**Context**:
The use case for this is torch.profiler for python subclasses.

Autograd attaches a "sequence number" to all ops that it encounters during the forward pass. Then, the corresponding sequence number can be associated with a backward kernel when backward is executed. This is used by the profiler to associate the forward ops to the backward ops; a forward op and a backward op with the same sequence number are "linked" in some post-processing step.

Prior to this PR, this profiler feature didn't work for python subclasses. The reason is that we don't collect profiler information for all the dispatches for a given kernel; we only dispatch the initial `call`, and not the subsequent `redispatch` invocations. Normally, an Autograd key (if we're running with autograd) is the highest dispatch key, so the initial `call` that we profile is an Autograd key, and we collect the sequence number. But when we're dealing with a python subclass, the first dispatch key is PythonTLSSnapshot, which eventually redispatches into Autograd. We don't record the Autograd sequence number in that case because we don't record redispatches.

To fix this, this PR adds a special case for the dispatcher sequence number logic: we _also_ attach the sequence number to a `call` if its dispatchKey is `PythonTLSSnapshot` _and_ the dispatchKeySet contains autograd. This assumes that (a) PythonTLSSnapshot will be the first dispatch key, and (b) we will be redispatching into Autograd later, not "call"-ing.

[ghstack-poisoned]

davidberard98 committed:
Update on "[Dispatcher] Collect autograd sequence numbers on PythonTLSSnapshot dispatch keys"


Fixes #121758

**TL;DR**: When profiling is turned on, the dispatcher will sometimes attach the autograd sequence number to the recorded profiler event. This PR expands the set of profiler events onto which we attach sequence numbers. In addition to the previous events, we will now also attach sequence numbers for an event which is (a) called through "call" (as opposed to "redispatch"), (b) has PythonTLSSnapshot as the current dispatch key, and (c) has an Autograd key in the dispatchKeySet.

**Context**:
The use case for this is torch.profiler for python subclasses.

Autograd attaches a "sequence number" to all ops that it encounters during the forward pass. Then, the corresponding sequence number can be associated with a backward kernel when backward is executed. This is used by the profiler to associate the forward ops to the backward ops; a forward op and a backward op with the same sequence number are "linked" in some post-processing step.

Prior to this PR, this profiler feature didn't work for python subclasses. The reason is that we don't collect profiler information for all the dispatches for a given kernel; we only dispatch the initial `call`, and not the subsequent `redispatch` invocations. Normally, an Autograd key (if we're running with autograd) is the highest dispatch key, so the initial `call` that we profile is an Autograd key, and we collect the sequence number. But when we're dealing with a python subclass, the first dispatch key is PythonTLSSnapshot, which eventually redispatches into Autograd. We don't record the Autograd sequence number in that case because we don't record redispatches.

To fix this, this PR adds a special case for the dispatcher sequence number logic: we _also_ attach the sequence number to a `call` if its dispatchKey is `PythonTLSSnapshot` _and_ the dispatchKeySet contains autograd. This assumes that (a) PythonTLSSnapshot will be the first dispatch key, and (b) we will be redispatching into Autograd later, not "call"-ing.

[ghstack-poisoned]

davidberard98 committed:
Update on "[Dispatcher] Collect autograd sequence numbers on PythonTLSSnapshot dispatch keys"


Fixes #121758

**TL;DR**: When profiling is turned on, the dispatcher will sometimes attach the autograd sequence number to the recorded profiler event. This PR expands the set of profiler events onto which we attach sequence numbers. In addition to the previous events, we will now also attach sequence numbers for an event which is (a) called through "call" (as opposed to "redispatch"), (b) has PythonTLSSnapshot as the current dispatch key, and (c) has an Autograd key in the dispatchKeySet.

**Context**:
The use case for this is torch.profiler for python subclasses.

Autograd attaches a "sequence number" to all ops that it encounters during the forward pass. Then, the corresponding sequence number can be associated with a backward kernel when backward is executed. This is used by the profiler to associate the forward ops to the backward ops; a forward op and a backward op with the same sequence number are "linked" in some post-processing step.

Prior to this PR, this profiler feature didn't work for python subclasses. The reason is that we don't collect profiler information for all the dispatches for a given kernel; we only dispatch the initial `call`, and not the subsequent `redispatch` invocations. Normally, an Autograd key (if we're running with autograd) is the highest dispatch key, so the initial `call` that we profile is an Autograd key, and we collect the sequence number. But when we're dealing with a python subclass, the first dispatch key is PythonTLSSnapshot, which eventually redispatches into Autograd. We don't record the Autograd sequence number in that case because we don't record redispatches.

To fix this, this PR adds a special case for the dispatcher sequence number logic: we _also_ attach the sequence number to a `call` if its dispatchKey is `PythonTLSSnapshot` _and_ the dispatchKeySet contains autograd. This assumes that (a) PythonTLSSnapshot will be the first dispatch key, and (b) we will be redispatching into Autograd later, not "call"-ing.

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55724190).

davidberard98 committed:
Update on "[Dispatcher] Collect autograd sequence numbers on PythonTLSSnapshot dispatch keys"


Fixes #121758

**TL;DR**: When profiling is turned on, the dispatcher will sometimes attach the autograd sequence number to the recorded profiler event. This PR expands the set of profiler events onto which we attach sequence numbers. In addition to the previous events, we will now also attach sequence numbers for an event which is (a) called through "call" (as opposed to "redispatch"), (b) has PythonTLSSnapshot as the current dispatch key, and (c) has an Autograd key in the dispatchKeySet.

**Context**:
The use case for this is torch.profiler for python subclasses.

Autograd attaches a "sequence number" to all ops that it encounters during the forward pass. Then, the corresponding sequence number can be associated with a backward kernel when backward is executed. This is used by the profiler to associate the forward ops to the backward ops; a forward op and a backward op with the same sequence number are "linked" in some post-processing step.

Prior to this PR, this profiler feature didn't work for python subclasses. The reason is that we don't collect profiler information for all the dispatches for a given kernel; we only dispatch the initial `call`, and not the subsequent `redispatch` invocations. Normally, an Autograd key (if we're running with autograd) is the highest dispatch key, so the initial `call` that we profile is an Autograd key, and we collect the sequence number. But when we're dealing with a python subclass, the first dispatch key is PythonTLSSnapshot, which eventually redispatches into Autograd. We don't record the Autograd sequence number in that case because we don't record redispatches.

To fix this, this PR adds a special case for the dispatcher sequence number logic: we _also_ attach the sequence number to a `call` if its dispatchKey is `PythonTLSSnapshot` _and_ the dispatchKeySet contains autograd. This assumes that (a) PythonTLSSnapshot will be the first dispatch key, and (b) we will be redispatching into Autograd later, not "call"-ing.

Differential Revision: [D55724190](https://our.internmc.facebook.com/intern/diff/D55724190)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55724190).

albanD commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -498,24 +498,44 @@ std::vector<OperatorName> Dispatcher::getRegistrationsForDispatchKey(c10::option
+  // sequence number for operators on python subclasses.
+  //
+  // This is the special case: if we're profiling a PythonTLSSnapshot call
+  // and we have an autograd dispatch key in the dispatchKeySet, then we'll
```
I'm curious why the logic here is not simplified to always get the seq_num unless autograd is in the exclude set?

davidberard98 replied:
@albanD that's a good point.

My goal was to only capture the sequence_nr once during the forward pass in profiler events. My guess was that this is a safe assumption for PythonTLSSnapshot, but that it might not be true in general (i.e. if we `call` PythonTLSSnapshot, then we won't later `call` again on or above Autograd).

But I guess maybe I should instead just always get the sequence_nr and then fix the profiler logic so that it handles duplicate sequence_nr in forward. Let me experiment to see if I can repro this case and try to fix it

albanD replied:
From what I recall, the profiler already has to handle duplicates since the current logic can actually trigger multiple peek() of the same number when CompositeImplicit ops are involved.

davidberard98 committed:
Update on "[Dispatcher] Collect autograd sequence numbers on PythonTLSSnapshot dispatch keys"


Fixes #121758

**TL;DR**: When profiling is turned on, the dispatcher will sometimes attach the autograd sequence number to the recorded profiler event. This PR expands the set of profiler events onto which we attach sequence numbers. In addition to the previous events, we will now also attach sequence numbers for an event which is (a) called through "call" (as opposed to "redispatch"), (b) has PythonTLSSnapshot as the current dispatch key, and (c) has an Autograd key in the dispatchKeySet.

**Context**:
The use case for this is torch.profiler for python subclasses.

Autograd attaches a "sequence number" to all ops that it encounters during the forward pass. Then, the corresponding sequence number can be associated with a backward kernel when backward is executed. This is used by the profiler to associate the forward ops to the backward ops; a forward op and a backward op with the same sequence number are "linked" in some post-processing step.

Prior to this PR, this profiler feature didn't work for python subclasses. The reason is that we don't collect profiler information for all the dispatches for a given kernel; we only dispatch the initial `call`, and not the subsequent `redispatch` invocations. Normally, an Autograd key (if we're running with autograd) is the highest dispatch key, so the initial `call` that we profile is an Autograd key, and we collect the sequence number. But when we're dealing with a python subclass, the first dispatch key is PythonTLSSnapshot, which eventually redispatches into Autograd. We don't record the Autograd sequence number in that case because we don't record redispatches.

To fix this, this PR adds a special case for the dispatcher sequence number logic: we _also_ attach the sequence number to a `call` if its dispatchKey is `PythonTLSSnapshot` _and_ the dispatchKeySet contains autograd. This assumes that (a) PythonTLSSnapshot will be the first dispatch key, and (b) we will be redispatching into Autograd later, not "call"-ing.

Differential Revision: [D55724190](https://our.internmc.facebook.com/intern/diff/D55724190)

cc ezyang msaroufim albanD

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55724190).

albanD commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -498,24 +498,38 @@ std::vector<OperatorName> Dispatcher::getRegistrationsForDispatchKey(c10::option
+  // `call` has profiler instrumentation, whereas `redispatch` doesn't.
+  // So usually, we'll collect a sequence number on the first call() if the
+  // dispatch keys contain autograd, and not on subsequent redispatches.
+  bool dispatchHasAutograd = !(dispatchKeySet & autograd_dispatch_keyset).empty();
```
btw my suggestion was to check the exclude set here.
The way autograd works is that the key is added to every single Tensor. So everything will always dispatch to autograd unless:
- Autograd key is in the exclude set
- Inference mode is running

Since inference mode implies no_grad mode (checked below), I think it is enough to check the tls exclude set here. That way you don't need to pass any argument in and can simplify the overall logic quite a bit.

davidberard98 replied:
ah I see, thanks! I figured we didn't need to check given we checked no_grad mode below, but your explanation makes it clear why checking the exclude set is better

davidberard98 replied:
@albanD - actually, I think the exclude set should already be handled:

from callBoxed, this gets the dispatch key set, which is eventually passed into the sequenceNumberForRunningRecordFunction function that I'm modifying in this PR:
https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/aten/src/ATen/core/dispatch/Dispatcher.h#L718

and getDispatchKeySetBoxed eventually calls computeDispatchKeySet:
https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L170

computeDispatchKeySet checks the exclude set:
https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L46

Based on this, do you think it's fine to leave it as is? (Is GradMode expected to show the exact same info as the include/exclude dispatch key sets? I checked and it wasn't obvious to me)

soulitzer replied:
gradmode and include/exclude can diverge for example (if this is what you are asking) because when you redispatch in the autograd kernel, you first disable autograd via the exclude key. If you later have a callBoxed  in torch dispatch, gradmode can be still enabled, but the autograd key has now been excluded.

soulitzer replied:
> actually, I think the exclude set should already be handle

I think the idea is that if checking only exclude set is sufficient to determine whether autograd kernel is reached, then we can avoid threading dispatch key at all these callsites and instead just check TLS inside sequenceNumberForRunningRecordFunction

However, if you create tensors within inference mode and then use them outside of inference mode, you will not reach the autograd kernel even though you are outside of no_grad, and autograd is not in TLS exclude set.

So, TLDR - the current appraoch is probably fine.

soulitzer reviewed (approved):
Thanks!

davidberard98 commented:
@pytorchbot merge
